SU2_CFD/src/solvers/CFEASolver.cpp

```diff
@@ -0,0 +1,3774 @@
+  unsigned long iVertex;
+
+  for (iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++){
+    switch (config->GetMarker_All_KindBC(iMarker)) {
+      case LOAD_BOUNDARY:
+        /*--- Only if the load is nonzero - reduces computational cost ---*/
+        if(config->GetLoad_Value(config->GetMarker_All_TagBound(iMarker)) != 0 ) {
+          /*--- For all the vertices in the marker iMarker ---*/
+          for (iVertex = 0; iVertex < geometry->nVertex[iMarker]; iVertex++) {
+            /*--- Retrieve the point ID ---*/
+            iPoint = geometry->vertex[iMarker][iVertex]->GetNode();
+            /*--- Clear the residual of the node, to avoid adding on previous values ---*/
+            nodes->Clear_SurfaceLoad_Res(iPoint);
+          }
+        }
+        break;
+      case DAMPER_BOUNDARY:
+        /*--- For all the vertices in the marker iMarker ---*/
+        for (iVertex = 0; iVertex < geometry->nVertex[iMarker]; iVertex++) {
+          /*--- Retrieve the point ID ---*/
+          iPoint = geometry->vertex[iMarker][iVertex]->GetNode();
+          /*--- Clear the residual of the node, to avoid adding on previous values ---*/
+          nodes->Clear_SurfaceLoad_Res(iPoint);
+        }
+        break;
+    }
+  }
+
+  /*
+   * FSI loads (computed upstream) need to be integrated if a nonconservative interpolation scheme is in use
+   */
+  if (fsi && first_iter && consistent_interpolation) Integrate_FSI_Loads(geometry,config);
+
+}
+
+void CFEASolver::SetInitialCondition(CGeometry **geometry, CSolver ***solver_container, CConfig *config, unsigned long TimeIter) {
+
+  /*--- We store the current solution as "Solution Old", for the case that we need to retrieve it ---*/
+
+  if (config->GetIncrementalLoad()) nodes->Set_OldSolution();
+
+}
+
+void CFEASolver::ResetInitialCondition(CGeometry **geometry, CSolver ***solver_container, CConfig *config, unsigned long TimeIter) {
+
+  /*--- We store the current solution as "Solution Old", for the case that we need to retrieve it ---*/
+
+  if (config->GetIncrementalLoad()) nodes->Set_Solution();
+
+}
+
+void CFEASolver::Compute_StiffMatrix(CGeometry *geometry, CNumerics **numerics, CConfig *config) {
+
+  const bool topology_mode = config->GetTopology_Optimization();
+  const su2double simp_exponent = config->GetSIMP_Exponent();
+  const su2double simp_minstiff = config->GetSIMP_MinStiffness();
+
+  /*--- Start OpenMP parallel region. ---*/
+
+  SU2_OMP_PARALLEL
+  {
+#ifdef HAVE_OMP
+    /*--- Chunk size is at least OMP_MIN_SIZE and a multiple of the color group size. ---*/
+    auto chunkSize = roundUpDiv(OMP_MIN_SIZE, ColorGroupSize)*ColorGroupSize;
+
+    /*--- Loop over element colors. ---*/
+    for (auto color : ElemColoring)
+    {
+      SU2_OMP_FOR_DYN(chunkSize)
+      for(auto k = 0ul; k < color.size; ++k) {
+
+        auto iElem = color.indices[k];
+#else
+    /*--- Natural coloring. ---*/
+    {
+      for (auto iElem = 0ul; iElem < nElement; iElem++) {
+#endif
```

Comment:
The coloring is then used like this and the body of the loop is almost unchanged apart from the use of local variables and picking one CElement and one CNumerics per thread.
When compiling without OpenMP support everything reverts to the natural way of looping over elements and no overhead or indirection is introduced by coloring.

Reply:
Is this linked to the number of threads of the preconditioners, or totally independent?

Reply:
Totally independent.